### PR TITLE
[FLINK-11519][table] Add function related catalog APIs

### PR DESCRIPTION
**Pull request by xuefuz**
## What is the purpose of the change

As a followup for FLINK-11474 and its PR#8007, this PR adds additional catalog APIs for functions (UDFs)

## Brief change log

   - Added APIs in ReadableCatalog and ReadableWritableCatalog
   - Provided an implementation of the APIs for GenericInMemoryCatalog
   - Supplied with corresponding tests

## Verifying this change

Manually tested and provided UTs

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by xuefuz** [comment]
This is to replace PR https://github.com/apache/flink/pull/8212. cc: @JingsongLi @KurtYoung

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableWritableCatalog.java [hunk 1]
```diff
@@ -195,4 +197,47 @@ void dropPartition(ObjectPath tablePath, CatalogPartitionSpec partitionSpec, boo
...
+	 * @throws CatalogException in case of any runtime exception
+	 */
+	void alterFunction(ObjectPath functionPath, CatalogFunction newFunction, boolean ignoreIfNotExists)
+		throws FunctionNotExistException, CatalogException;
+
+	/**
+	 * Drop a function.
+	 *
+	 * @param functionPath       path of the function to be dropped
+	 * @param ignoreIfNotExists  plag to specify behavior if the database does not exist:
```
Comment (KurtYoung) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/catalog/ReadableWritableCatalog.java]: database -> function

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java [hunk 2]
```diff
@@ -287,7 +292,84 @@ public CatalogBaseTable getTable(ObjectPath tablePath) throws TableNotExistExcep
...
+		checkNotNull(tablePath);
+
+		return databaseExists(tablePath.getDatabaseName()) && tables.containsKey(tablePath);
+	}
+
+	// ------ functions ------
+
+	@Override
+	public void createFunction(ObjectPath functionPath, CatalogFunction function, boolean ignoreIfExists)
+		throws FunctionAlreadyExistException, DatabaseNotExistException {
```
Comment (KurtYoung) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java]: indent with 2 tabs

Comment (xuefuz) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/GenericInMemoryCatalog.java]: I checked and it's two tables.

**Comment by xuefuz** [comment]
Thanks for Kurt's detailed review. I have updated the PR to reflect Kurt's suggestions. Please take another look and let me know if additional changes are needed. Thanks.

**Pull request merged**